### PR TITLE
Fix static-lib AOT Mono builds

### LINKED ISSUE #103982: Runtime tests fail to AOT compile on Apple mobile platforms
## Build Information
Build: https://dev.azure.com/dnceng-public/public/_build/results?buildId=719424
Build error leg or test failing: https://dev.azure.com/dnceng-public/public/_build/results?buildId=719424&view=logs&j=da83f4c3-af55-5024-34f6-3ac8d11ac15f

## Error Message

Fill the error message using [step by step known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssueJsonStepByStep.md).

```json
{
  "ErrorMessage": "duplicate symbol '_dn_simdhash_ptrpair_ptr_meta'",
  "BuildRetry": false,
  "ExcludeConsoleLog": false
}
```

```
...
[/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018: duplicate symbol '_dn_simdhash_ptrpair_ptr_vtable' in: [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018:     /tmp/helix/working/95610881/w/A2B908C5/e/publish/libmono-component-diagnostics_tracing-static.a(dn-simdhash-ptrpair-ptr.c.o) [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018:     /tmp/helix/working/95610881/w/A2B908C5/e/publish/libmonosgen-2.0.a(dn-simdhash-ptrpair-ptr.c.o) [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018: duplicate symbol '_dn_simdhash_ptrpair_ptr_try_add' in: [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018:     /tmp/helix/working/95610881/w/A2B908C5/e/publish/libmono-component-diagnostics_tracing-static.a(dn-simdhash-ptrpair-ptr.c.o) [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018:     /tmp/helix/working/95610881/w/A2B908C5/e/publish/libmonosgen-2.0.a(dn-simdhash-ptrpair-ptr.c.o) [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018: duplicate symbol '_dn_simdhash_ptrpair_ptr_meta' in: [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018:     /tmp/helix/working/95610881/w/A2B908C5/e/publish/libmono-component-diagnostics_tracing-static.a(dn-simdhash-ptrpair-ptr.c.o) [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018:     /tmp/helix/working/95610881/w/A2B908C5/e/publish/libmonosgen-2.0.a(dn-simdhash-ptrpair-ptr.c.o) [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018: ld: 156 duplicate symbols for architecture arm64 [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
```

The regression is likely introduced in https://github.com/dotnet/runtime/pull/103184.

/cc: @jkoritzinsky 

 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=719424
**Error message validated:** `[duplicate symbol '_dn_simdhash_ptrpair_ptr_meta'`]
**Result validation:** :white_check_mark: Known issue matched with the provided build.
**Validation performed at:** 6/25/2024 6:36:10 PM UTC

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[722476](https://dev.azure.com/dnceng-public/public/_build/results?buildId=722476)|dotnet/runtime|[baseservices_TieredCompilation.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=722476&view=ms.vss-test-web.build-test-results-tab&runId=18129490&resultId=100013)||
|[721722](https://dev.azure.com/dnceng-public/public/_build/results?buildId=721722)|dotnet/runtime|[baseservices_TieredCompilation.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=721722&view=ms.vss-test-web.build-test-results-tab&runId=18107258&resultId=100012)||
|[720898](https://dev.azure.com/dnceng-public/public/_build/results?buildId=720898)|dotnet/runtime|[baseservices_TieredCompilation.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=720898&view=ms.vss-test-web.build-test-results-tab&runId=18081448&resultId=100013)||
|[720816](https://dev.azure.com/dnceng-public/public/_build/results?buildId=720816)|dotnet/runtime|[baseservices_TieredCompilation.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=720816&view=ms.vss-test-web.build-test-results-tab&runId=18078286&resultId=100013)||
|[719424](https://dev.azure.com/dnceng-public/public/_build/results?buildId=719424)|dotnet/runtime|[baseservices_TieredCompilation.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=719424&view=ms.vss-test-web.build-test-results-tab&runId=18033732&resultId=100013)||
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|0|5|5|

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Fixes https://github.com/dotnet/runtime/issues/103982

### CODE REVIEW

**Comment by jkoritzinsky** [comment]
@kotlarmilos what pipeline should I run to ensure I'm getting good coverage here?

**Comment by kotlarmilos** [comment]
> @kotlarmilos what pipeline should I run to ensure I'm getting good coverage here?

`*_RuntimeTests` should provide good coverage for this fix. Please ignore the library tests on Apple mobile as this is an ongoing issue.

**Comment by jkoritzinsky** [comment]
Looks like it's working. I'll mark this ready for review.

**Comment by kotlarmilos** [comment]
@jkoritzinsky Please take a look at simulator failures as they seem related: https://dev.azure.com/dnceng-public/public/_build/results?buildId=721222&view=logs&j=07effcfd-c63d-5008-e811-03b29530f5cb&t=1b65cd35-7643-520a-edf8-09ec895c1ca1

File: src/mono/mono/component/CMakeLists.txt [hunk 1]
```diff
@@ -231,7 +231,9 @@ if (ENABLE_PERFTRACING AND "${MONO_DIAGNOSTICS_TRACING_COMPONENT_NAME}" IN_LIST
         list(APPEND mono-components-objects $<TARGET_OBJECTS:eventpipe-mono-objects> $<TARGET_OBJECTS:dn-containers>)
     endif()
     if (TARGET mono-component-${MONO_DIAGNOSTICS_TRACING_COMPONENT_NAME}-static)
-        target_sources(mono-component-${MONO_DIAGNOSTICS_TRACING_COMPONENT_NAME}-static PRIVATE $<TARGET_OBJECTS:eventpipe-mono-objects> $<TARGET_OBJECTS:dn-containers>)
+        # We don't add dn-containers here as any scenario that uses this target will also use the monosgen-static target,
```
Comment (steveisok) [inline src/mono/mono/component/CMakeLists.txt:234]: @akoeplinger do you recall if that's true even for simulator?

Comment (akoeplinger) [reply src/mono/mono/component/CMakeLists.txt:234]: I think so, yes

**Comment by kotlarmilos** [comment]
This fix resolves the device failure caused by symbol duplication. However, simulators are still encountering a different issue. Looking at logs between device and simulator, I noticed the following difference:

https://github.com/dotnet/runtime/blob/9906682033b4fee6aadc12ac878ffa8fde1edfe1/src/mono/mono.proj#L72-L75 

https://github.com/dotnet/runtime/blob/9906682033b4fee6aadc12ac878ffa8fde1edfe1/src/mono/mono.proj#L682

@akoeplinger Could we align these params for the device and simulator?

**Comment by jkoritzinsky** [comment]
The simulator failures are only when building the `Mono.debug.framework` and `Mono.release.framework` frameworks.

I'm doing my best to investigate, but I don't have a Mac available today.

**Review by kotlarmilos — APPROVED** [review]

**Comment by jkoritzinsky** [comment]
/ba-g Failure is an AzDO infra issue

**Pull request merged**